backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/serialization/KonanIrModuleDeserializer.kt

```diff
@@ -0,0 +1,327 @@
+        ) {
+            forwardDeclarations.add(symbol)
+        }
+
+        return symbol
+    }
+
+    override fun deserializeDescriptorReference(proto: KonanIr.DescriptorReference)
+        = descriptorReferenceDeserializer.deserializeDescriptorReference(proto)
+
+    private val ByteArray.codedInputStream: org.jetbrains.kotlin.protobuf.CodedInputStream
+        get() {
+            val codedInputStream = org.jetbrains.kotlin.protobuf.CodedInputStream.newInstance(this)
+            codedInputStream.setRecursionLimit(65535) // The default 64 is blatantly not enough for IR.
+            return codedInputStream
+        }
+
+    private val reversedFileIndex = mutableMapOf<UniqIdKey, IrFile>()
+
+    private val UniqIdKey.moduleOfOrigin get() =
+        this.moduleDescriptor ?: reversedFileIndex[this]?.packageFragmentDescriptor?.containingDeclaration
+
+    private fun deserializeTopLevelDeclaration(uniqIdKey: UniqIdKey): IrDeclaration {
+        val proto = loadTopLevelDeclarationProto(uniqIdKey)
+        return deserializeDeclaration(proto, reversedFileIndex[uniqIdKey]!!)
+    }
+    
+    private fun reader(moduleDescriptor: ModuleDescriptor, uniqId: UniqId) = moduleDescriptor.konanLibrary!!.irDeclaration(uniqId.index, uniqId.isLocal)
+
+    private fun loadTopLevelDeclarationProto(uniqIdKey: UniqIdKey): KonanIr.IrDeclaration {
+        val stream = reader(uniqIdKey.moduleOfOrigin!!, uniqIdKey.uniqId).codedInputStream
+        return KonanIr.IrDeclaration.parseFrom(stream, KonanSerializerProtocol.extensionRegistry)
+    }
+
+    private fun findDeserializedDeclarationForDescriptor(descriptor: DeclarationDescriptor): DeclarationDescriptor? {
+        val topLevelDescriptor = descriptor.findTopLevelDescriptor()
+
+        if (topLevelDescriptor.module.isForwardDeclarationModule) return null
+
+        if (topLevelDescriptor !is DeserializedClassDescriptor && topLevelDescriptor !is DeserializedCallableMemberDescriptor) {
+            return null
+        }
+
+        val descriptorUniqId = topLevelDescriptor.getUniqId()
+            ?: error("could not get descriptor uniq id for $topLevelDescriptor")
+        val uniqId = UniqId(descriptorUniqId.index, isLocal = false)
+        val topLevelKey = UniqIdKey(topLevelDescriptor.module, uniqId)
+
+        reachableTopLevels.add(topLevelKey)
+
+        // TODO: This is a mess. Cleanup!
+        do {
+            val key = reachableTopLevels.first()
+
+            if (deserializedSymbols[key]?.isBound == true) {
+                reachableTopLevels.remove(key)
+                continue
+            }
+
+            val previousModuleDescriptor = deserializedModuleDescriptor
+            deserializedModuleDescriptor = key.moduleOfOrigin
+
+            if (deserializedModuleDescriptor == null) {
+                deserializedModuleDescriptor = previousModuleDescriptor
+                reachableTopLevels.remove(key)
+                deserializedTopLevels.add(key)
```

Comment:
Is this correct? What case does this code handle?

Reply:
Cleaned up a little.